### ISSUE TITLE
Show codepoint under cursor

### ISSUE THREAD

**Issue by ygoe** [notepad-plus-plus/notepad-plus-plus#2056]
It would be really helpful sometimes to know what character is at the cursor. What you see is the visual glyph presentation but that's not unique. There are a dozen of spaces, several hyphens, and other characters of the same appearance. To analyse any issues with same-looking but wrong characters, the Unicode codepoint could be displayed in the status bar. BabelPad does that for example, together with the character's name for easier understanding.

**Comment by MetaChuh** [comment]
please update your notepad++ to the latest version [<kbd>Notepad++ 7.6.6</kbd>](https://notepad-plus-plus.org/download/v7.6.6.html) and migrate or reinstall your plugins if you update from version 7.5.9 or below.
if your issue still persists, and it is still of relevance, please feel free to file a new issue as easy to understand and reproducible as possible.

thank you for your help and best regards.

**Comment by ygoe** [comment]
@MetaChuh Looks like you haven't even read what you just blindly closed, as you always do. And it is very disappointing, as a community member, to see suggestions simply ignored, and then ignored again, without even a single word in response. I consider this action an explicit statement that the Notepad++ team* is not interested in user feedback. And when it does exist, it is not handled in an appropriate way, i.e. either assigning it a target milestone (be it as far in the future as it may) or rejecting the suggestion with a relevant explanation why this is not desired. Yes, it does take a moment. But so does community interaction, fixing bugs and development in general. And you usually don't process thousands of forgotten issues years after. I'm wondering why this issue tracker even still exists.

As my last and final suggestion to this project, I suggest you to close this issue tracker down completely, as a clear and simple statement of what you think of the issue tracker and the people who write stuff in it. And it would save you from the hard work of mass-closing all issues with a stupid template text. Should I once forget about my intention and open any new issue here, I'm sure you'll remind me about it a few years later, as you always do.

I guess I'm not the only contributor you lost here.

*) I haven't looked up what your position in this project is, but from your permission to close others' issues, I infer that you are part of the team that makes the decisions. Please correct me if that's a false assumption.

**Comment by DerekRead** [comment]
Well, this is disappointing as I was hoping for such a feature. Having to run files through JavaScript to identify codepoints is annoying. Having the same feature directly in the Notepad++ UI would be so much nicer.

It looks like this feature has been requested twice now and rejected both times without explanation (this one and #1343). It seems like it would at least make sense to explain why a requested feature is being rejected.

**Comment by alankilborn** [comment]
@DerekRead and others:

Perhaps not ideal, but this is something an add-on script (a PythonScript) can do for you.
As an example of such a script, here's `CodepointsAndBytesOfUtf8CharacterAtCaret.py` :

    # -*- coding: utf-8 -*-
    from __future__ import print_function

    from Npp import *
    import inspect
    import os
    import sys

    #-------------------------------------------------------------------------------

    # some test data:
    #  four-byte: 𒊹
    #  three-byte: ₵
    #  two-byte: ¶
    #  misc: 💙 🡢 ⮃ 🠧 🠉 … ⮞ ⮜ ⮝ ⮟ ⮘ ⮚ ⮙ ⮛ ➢ ➣ ➤ 👍 café

    #-------------------------------------------------------------------------------

    class CABOU8CAC(object):

        def __init__(self):
            self.this_script_name = inspect.getframeinfo(inspect.currentframe()).filename.split(os.sep)[-1].rsplit('.', 1)[0]
            python2 = True if sys.version_info.major == 2 else False
            self.debug = True if 0 else False
            p = editor.getCurrentPos()
            q = editor.positionAfter(p)
            char_size_in_bytes = q - p
            output = ''
            if char_size_in_bytes == 0:
                output =  'end-of-file'
            else:
                if self.debug: print('char_size_in_bytes:', char_size_in_bytes)
                c = editor.getTextRange(p, q)
                if char_size_in_bytes == 1:
                    if c == '\r': output = 'line-ending: CR'
                    elif c == '\n': output = 'line-ending: LF'
                    elif c == '\t': output = 'character: TAB'
                    elif c == ' ': output = 'character: SPACE'
                    else: output = 'character: ' + c
                elif char_size_in_bytes == 2 and c[0] == '\r' and c[1] == '\n':
                    output = 'atomic-line-ending: CRLF'
                else:
                    # under Python2, c is a str(bytestr); under Python3, c is a str(unicode)
                    c_bytestr = c_unistr = c
                    if python2:
                        c_unistr = c.decode('utf-8')
                    else:
                        c_bytestr = c.encode('utf-8')
                    output = 'codepoint: {cp}\r\nbytes: {b}'.format(
                        cp=c_unistr.encode('unicode_escape'), b=repr(c_bytestr) if python2 else c_bytestr)
            if len(output): self.mb(output)

        def mb(self, msg, flags=0, title=''):  # a message-box function
            return notepad.messageBox(msg, title if title else self.this_script_name, flags)

    #-------------------------------------------------------------------------------

    if __name__ == '__main__': CABOU8CAC()

You put your caret on the character you want to analyze, run the script, and a box like this pops up:

![image](https://user-images.githubusercontent.com/77065706/159566756-97a43a18-06f0-4439-b1a7-432000ef8acb.png)

This is based on some earlier Community postings for similar wishes:

* https://community.notepad-plus-plus.org/topic/17799/how-to-see-hex-value-of-character-next-to-cursor
* https://community.notepad-plus-plus.org/topic/14767/ascii-hex-of-current-character

**Comment by DerekRead** [comment]
@alankilborn Thanks for the script. This might be good enough for me. I've been using other tools just for this at the moment so I'll have to weigh that against messing around with scripting in Notepad++.

If Notepad++ provides access to writing to its status then putting the result from your script in there would probably make everyone happy. I assume not though, because you might have done so if it is possible.

I wrote something similar to yours in JavaScript for the software I used to work on (XMetaL) and it was used for a few years by customers before we finally turned it into a first class feature. The difference there was that XMetaL had an API to write to the statusbar so my script actually did what @ygoe wants in Notepad++. Getting the value was trivial (similar to your script using JavaScript charCodeAt) with the limitation that it only covered U+0000 to U+FFFF. I had a lookup for all the Unicode character names at one point too but that took up too much space in the statusbar, and the Unicode database needed for that was huge (everything except Korean which I generated based on the formula for that, and I ignored Chinese as it is a complex mess in a separate database). Just providing the U+#### value is good enough for me as that is easy to lookup if I don't already know it.

I'll poke around in those other script examples as well. Perhaps there's something there.

**Comment by alankilborn** [comment]
@DerekRead There is access to the status bar; I didn't do it in the script posted here because for my use it is fine as an on-demand thing.  Those scripts at the Community site will show how to write to the status bar.

**Comment by DerekRead** [comment]
@alankilborn Thanks.
I have this running now. Didn't realise a plugin for Python was required but didn't take much to figure that out.

I think this is good enough for my immediate needs. Not super user-friendly but at least I now have it running from a button.
Doesn't look like there's a way to configure a shortcut key.

I think if I was going to extend this, while keeping the dialog, I might have it display info for the first 20 or something characters from my selection. That's close to what I need for examining files. Should let me idenfity some wonky portion of a document well enough before I need to get serious and hunt around in a hex editor.

If I was to have it display something in status I'd want that to run whenever the selection changes. Not sure if Notepad++ has events that can trigger scripts. I don't think I have the time to look into that, and it seems like that would be best left for a first class feature anyway.

**Comment by DerekRead** [comment]
So, there is no location on the statusbar to insert a string without clobbering existing info:
https://npp-user-manual.org/docs/plugin-communication/#nppm-setstatusbar

That's unfortunate, but it makes sense. It took me 2 years at my previous company to get dev to add an additional, normally blank, statusbar section that customizers could use to display their own info.

I guess I could just clobber one of the existing bits of info (STATUSBAR_TYPING_MODE is mostly useless to me).
Grabbing an existing status string, appending to it and writing it all back is an idea too. Obviously whatever the current value is needs to be accessible, and I still have yet to find "events" in the docs that will let me run a script on selection change.

**Comment by DerekRead** [comment]
@alankilborn I've got your script mapped to Ctrl+Alt+C. I don't like the exact details shown for my particular usage but I won't bother posting anything else here as I think my implementation is going to be pretty specific to work around the limitations of how this is displayed in a dialog rather than statusbar. -- Thanks again.

I'm still +1 for showing the codepoint value (in U+#### format) in the statusbar for the next character in the document (as a first class feature). To me it looks like it really belongs in there with "Ln: #  Col: #  Pos: #" and would only extend that by 6 or 7 characters:
`Ln: #  Col: #  Pos: #  U+####`
or `Ln: #  Col: #  Pos: #  U+#####`

![notepad++ feature request](https://user-images.githubusercontent.com/19498041/159606204-d059db15-13f9-46fa-8b00-5e7e2424b0e0.png)

**Comment by alankilborn** [comment]
>Doesn't look like there's a way to configure a shortcut key.

Yes there is.  The basics of how are [HERE](https://community.notepad-plus-plus.org/post/72575).

>Not sure if Notepad++ has events that can trigger scripts.

Not whole scripts, but functions (called "callbacks") which you define within a script file.

>If I was to have it display something in status I'd want that to run whenever the selection changes

I think the arguments to the `SCINTILLANOTIFICATION.UPDATEUI` callback will provide such selection information.

>...and it seems like that would be best left for a first class feature anyway.

Sure, I guess, but what are the odds this kind of thing will ever appear as a "first class feature"?
Just look at this issue; opened July of 2016...coming up on 6 years...

I've seen so many times where people make a feature request, and I propose a workaround that IMO provides excellent functionality, and the reaction I get is _That's nice and thank you...but I think I'll just wait for the 'first class feature' to appear in Notepad++_.
99% of those people are still waiting, when they could have their feature working immediately (in an add-on form) -- it boggles the mind.

**Comment by alankilborn** [comment]
@DerekRead 

>I think my implementation is going to be pretty specific to work around the limitations of how this is displayed in a dialog rather than statusbar.

Another option is to keep the PythonScript console window open but small in vertical height.  Then a `print()` statement in a script will cause output to appear there.

**Comment by DerekRead** [comment]
I found the callback stuff and have it updating the status after looking at some of those other posts.

That isn't something a lot (probably most) users of Notepad++ will be able to muddle their way through and get working though, and for those users just having this feature work out of the box might be nice.

**Comment by alankilborn** [comment]
>That isn't something a lot (probably most) users of Notepad++ will be able to muddle their way through and get working though

I gave pretty much a turnkey solution [HERE](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/2056#issuecomment-1075590747).  If you didn't like that exact solution, well, that's going to be kind of a problem for ANY solution, even a native one.  I for one am glad that I have the possibility (and the ability) to create solutions that work for me.  Of course when I share them, I also hope others get good results from them.  :-)

>and for those users just having this feature work out of the box might be nice.

I totally agree.  But N++ devs have to pick-n-choose what they have time to implement.

**Comment by donho** [comment]
I will see what I can do about it.

**Comment by DerekRead** [comment]
Thanks @donho

@alankilborn
Yep, totally agree with all your points (I've been saying this to people for 20+ years now: "devs have to pick-n-choose what they have time to implement").

I've got a working system that basically does what I need so this request is purely for other users at this point. That I have the feature doesnt negate the point that many users of N++ would probably not be able to figure that out. It also seems like something that would be fairly trivial for the right dev to implement in the core product, but that's just a guess, and it seems to have minimal impact on the UI which is also a consideration. If this pops up as a feature one day it won't get in the way of anyone that doesn't need it even if it was always on, or if they enable it by mistake.

**Comment by alankilborn** [comment]
Parting thought:  One thing about add-on features is that they can be very specific to a need, and don't need to consider all aspects of a feature and all users.  When it is a native feature, it DOES have to do that.  And there can be a bit of work (design and actual) to that.  UTF-8 isn't the only encoding in the world, and indeed Notepad++ supports many.

But...  I guess so much talk here about this one feature got the devs attention! :-)

**Comment by alankilborn** [comment]
@leonbloy:
The GotoLineCol plugin offers such a feature.
![image](https://user-images.githubusercontent.com/77065706/206004145-14003002-11f7-4d17-881e-e6e92d8ddaba.png)

**Comment by leonbloy** [comment]
@alankilborn Thanks.

**Comment by ygoe** [comment]
> More char info than I need but too much information is never bad as long as what I do need is in there somewhere.

Disagree. It's too complicated to get there, then it takes too much screen space, and extra work to free that space again. The plugin might be good for its primary task (I had installed it but don't need it for that) but not for this simple display.

Thanks for the VSCode extension, I've already installed that. It's much closer to what I want.

**Comment by DerekRead** [comment]
Yeah, it does take up some space, but I have it floating so not too horrible for me. This is much better than anything I have tried to date as once I have it open Notepad++ remembers that between sessions so it's always there for me, and the fact that it updates as soon as I move the cursor is the main thing I wanted.

Ideally it would still be nice to have something that just puts the char codepoint value in the statusbar or somewhere else where it is out of the way. In the dedicated XML editors I use (a few different tools) that's where it typically sits (for those that have this).

![image](https://user-images.githubusercontent.com/19498041/206049735-8e1640ed-5581-4501-b114-ab5b19b02e2b.png)

![image](https://user-images.githubusercontent.com/19498041/206049864-ff624217-62d0-4e38-85ee-93e5fb97392f.png)

**Comment by another-git-hub-user** [comment]
I've been use [PSPad](https://www.pspad.com/) because of this exact function since the late 90's! Still find NP++ excellent too!

**Comment by pryrt** [comment]
There is now a plugin called [Emoji Description](https://github.com/Ruberoid/npp_emoji_description), which replaces the "file type" field of the status bar to provide information on the current character (whether it is emoji or not):

<img width="441" height="32" alt="Image" src="https://github.com/user-attachments/assets/ea1d3250-b971-4fbd-90f7-a937a212097b" />

You can just toggle on/off the plugin's **Plugin > Emoji Description > Show Character Info** toggle action in order to switch between the character info and Notepad++'s default use of the file-type field (you may have to change tabs away from and back to the active tab to trigger Notepad++ to update back to the default value).

So with Emoji Description or GotoLineColumn plugins, you can either get the information on the status bar or in a docking panel, whichever makes the most sense for you.
